### DISCUSSION TITLE
MC-LSTM Application on short-term forecasting

### DISCUSSION THREAD

**Discussion by arasburakaslan** [neuralhydrology/neuralhydrology#84] [Q&A] [answered]
Hello Kratzert,

I am now working on my thesis about the application of physics-informed deep learning models to forecasting flash floods. MC-LSTM is one of the algorithms I'm using, and I'm attempting to design my model to operate with neuralhydrology. Along with the setups, I have a few application-related questions.

I'm attempting to forecast discharges six hours in advance (Qt+6) using 48 timesteps of rainfall and discharge data [Qt-48:Qt] and [Pt-48:Pt]. So my sample size, in other terms sequence length, is 48 and the forecast step is 6. My questions are as follows:

1.To begin, the configuration file (.yml) lacks a section for MC-LSTM. I'm considering using both discharge and rainfall as mass inputs because they both contribute to the total mass. Is this possible with neuralhydrology, I tried but it was saying that it only allows one mass input? Do you think it will be easier to create a model from scratch using mc-lstm repository other than arranging the neuralhydrology for my specific purpose?

2. I am not using any forcing data and am instead relying on dynamic hourly data. Does neuralhydrology allow the usage of hourly data or its only for daily predictions?

3. Because I am using a single catchment, my validation and test sets are combined into a single time series. As I will not be validating with a random basin, I set this value to empty. Is this valid?

4. My inputs are precipitation (mm/hr) and discharge (mm/hr), and my output is only the discharge. I employ sliding windows for sampling in a typical LSTM model, but I'm not sure if that's possible here. There is a lagged feature argument that I couldn't locate in the config.yml file and couldn't figure out how to utilize. To model Qt+6, I'm considering adding a new column to my dataset called Qt+6 and using it only for target values. Thus, for a sample, my inputs will be hourly P and Q values for the previous 48 hours and my output will be Qt+6. In that case, what should the predicted last n value be? Indeed, I was confused by this argument based on the definitions. If the length of my sequence is 48 and the predicted _last n is 6, would this give me the loss for prediction on the 54th or 42nd hour which I already have?

5. Does the model updates its gradients after each batch or each epoch?¨

Thank you in advance for answering my questions.

**Comment by kratzert** [comment, marked as answer]
Hi @arasburakaslan I think there might be some confusion/misconception on various points. Let me try to answer your questions point-by-point:

1. You need to write that config file (it is a YAML file that you can write with any editor, such as Visual Studio Code). You can start with any config from the tutorials and add/remove/change any argument that you like. If you are missing an argument in these config examples that you see listed in https://neuralhydrology.readthedocs.io/en/latest/usage/config.html simply add it yourself.
2. I don't understand your question. You are not using forcing data, but precipitation is a forcing, no? And regarding temporal resolution: All models can be used with any temporal resolution and we even have a model that can work with different temporal resolutions at the same time (see MTS-LSTM).
3. This is your choice however let me state clearly that this is not best-practice: a) you want to have a 3-fold split to have one split for hyper parameter tuning and one split for the independent model evaluation (aka test set). The concept of train/val/test has nothing to do with single basin oder multiple basins.
4. You should have another look at the MC-LSTM and yes, for your use-case it is not defined. You are trying to mass-balance discharge + precipitation as inputs vs discharge as output. If you want to use the MC-LSTM (as with any mass conserving hydrology model), your mass inputs should be "mass added to the system" (i.e. precipitation) and your mass output will be "mass removed from the system" (i.e. discharge). An autoregressive model as you describe it is not what the MC-LSTM is for. Instead, you probably want to use the  standard LSTM (here implemented under `cudalstm`).
5. After each batch

All your questions make me belief that you should probably spend some time to get familiar with this library by e.g. following all the tutorials first and trying to play around with the code. I hope my answers do help to clarify some of your questions.

**Comment by mengziwen2000** [comment]
I'd like to see the settings for the "trash cell", but I don't see the detailed setup of the "trash cell" in the mclstm.py file, it just removes the variable for the first hidden state. 
```
m_out, c = self.mclstm(x_m, x_a)
output = m_out[:, :, 1:].sum(dim=-1, keepdim=True)
```
 So how does the first hidden state correspond to evapotranspiration? For regionalised large models, considering ET is indeed a necessary option, but for special basins, such as karst landscapes (karstic landscapes), then the groundwater quantity is also used as a rubbish unit. Maybe it is possible to show the power of the MC-LSTM model in terms of forecasting accuracy in particular areas. This is what I thought during the process of reviewing the paper in recent times, and maybe this is incorrect.

**Reply by kratzert** [reply]
Hi, in our setup there was really not anything more sophisticated than to provide the model with some means to remove mass from the system in other ways than streamflow output. So we only implemented the hacky trick that you found and there was really nothing more to it. What we did post-training was to analyze the timeseries of the data that is being removed by the system in this first mass output (which is ignored for the prediction). Here, we could see that the signal correlates quite well with evaporation. That being said, I would not argue at this point that it actually _is_ evapotranspiration. And yes, in areas with a lot of karst, this might be a combination of evapotranspiration and water being "lost" through the cave system. I didn't spend much time thinking about this now and it is not really clear to me how you want to add two cells that should model two separate "rubbish" processes (evapotranspiration and water lost in karst system), without any target for either of the two.

**Reply by mengziwen2000** [reply]
Thank you for your reply. I was also trying to figure out how to add two units to simulate two separate "rubbish" processes. This will only wait until doing some work related to hydrological forecasting in karst areas and using your MC-LSTM to do some experiments.

**Comment by Iftekhar273** [comment]
Hi can you provide me your dataset? I am also working on flash flood prediction
